meltAssay

microsud opened the pull request:
Tested on windows 10, R 4.0.3 with `devtools::check()`. No errors and warnings.

microsud committed:
meltAssay

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,135 @@
+  signature = c(x = "MicrobiomeExperiment"),
+  function(x, colData_choice = "all", abund_values = "counts") {
+    # input check
+    if (isFALSE(any(abund_values == c("counts", "relabundance", "both")))) {
```
You can checkout the function `.check_abund_values` in `MicrobiomeExperiment`. I added that as the universal check for a certain assay to be present. I also don't like to hard code assay names. How the user names their assay and maybe wants to compare two assays of the same type, is up to them.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,135 @@
+
+      dfSE <- me_count_ab %>%
+        left_join(me_rel_ab, by = c("uTaxaID", "uSamId"))
+    }
```
This whole section has to be split up into different functions. One function should to one thing.

`meltAssay` should melt only one assay. The used can then merge how many assays as the like. Currently only one or two are allowed, which is too specific.

If you really need the this function, construct a second method which wraps this functionality, eg.  `meltAssays`.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,135 @@
+        mutate_if(is.factor, as.character)
+
+      return(dfSE)
+    }
```
I don't see the necessity for distinguishing `all` and `some`. technically is doesn't make a difference, which is shown by the fact that lines 115-117 and 128-130 are duplicates.

I would suggest, selecting the colData to use based on the chosen colums and then `left_join` them independent on how many there are.

In addition, one usage case is missing, which is `none`. I would argue, this should also be the default defined by the value `NULL`

Furthermore, `colData_choice` is mixing snake and camel case. I would strongly advise to revised this. Maybe `add_sample_data` or `add_col_data` would be a better name. Then the whole function could be formulated as a sentence, which is sometimes a good sanity check for naming arguments. The sentence `melt the assay with the abund_values "counts" and add the sample data "x,y and z" ` works. With colData_choice the sentence gets kind of clunky.

FelixErnst reviewed (changes requested):
Hi Sudarshan @microsud ,

thanks for the PR. I have a few comments on specific points in the code. See below.

In addition tests are missing. tests might benefit from splitting the separate steps into internal functions. For example I could see this split into `.melt_assay`, `.add_taxonomic_data_to_molten_assay` and `.add_col_data_to_molten_assay`. You can then write a little unit test for each internal function and one for the exported function at the end. This way future debugging gets much easier.

Thanks for considering making the changes.
Felix

microsud committed:
more checks and tests

microsud commented:
Attempt at addressing the comments on code and unit tests

FelixErnst commented:
Did you add tests in `tests/testthat`? I don't see any files added in that directory

microsud committed:
tests

microsud commented:
not sure why it was not added. Hopefully should be there now

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#' @param ... optional arguments not used.
+#' @return A data.frame tbl_df.
+#' @examples
+#' \dontrun{
```
Please set examples to be run. If we submit it to bioconductor they have to be run anyway

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#' @importFrom dplyr left_join mutate_if %>%
+#' @keywords Utilities
+#' @export
+setGeneric("meltAssay",
```
Have a look at some other files regarding the structuring of roxygen blocks. Sometimes it is helpful to structure the blocks with spaces and let the end in `NULL`. This way the man page is set up independent of any function. This is probably just a bit of personal flavour. Maybe you want to have look at other packages, like `scater` or `TreeSummarizedExperiment`.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+
+    uTaxaID <- uSamId <- Abundance <- NULL
+
+    dfSE <- .melt_assay(x, abund_values)
```
Alsways creating a new object is inefficient, since it increase RAM consumption. You are not reusing any of the objects, so just overwrite them

In addition if `add_col_data` would be `NULL` the function `add_col_data_to_molten_assay` could be skipped. This is what it could look like:

```
if(!is.null(add_col_data)){
    if(is.logical(add_col_data) && length(add_col_data) == 1L && add_col_data){
        add_col_data <- colnames(colData(x))
    } else {
        # check that all values are valid colnames. Otherwise through error
    }
    dfSE <- .add_col_data_to_molten_assay(dfSE, x, add_col_data)
}
```

The same principle would apply to `add_row_data`/`add_row_data_to_molten_assay`

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+    # input check
+    .check_abund_values(abund_values, x)
+
+    uTaxaID <- uSamId <- Abundance <- NULL
```
Not used.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+  uTaxaID <- NULL
+
+  # input check
+  .check_abund_values(abund_values, x)
```
You already checked above. So this is not necessary.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#' @param abund_values Must be one of the values returned by assayNames.
+#'
+.melt_assay <- function(x, abund_values) {
+  uTaxaID <- NULL
```
I am not sure, but you are not using this either. If using `!uTaxaID` below gets you an WARNING during R CMD check, try to solve it via `.data`. For details have a look at `?.data`

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#'          object.
+#'
+.add_taxonomic_data_to_molten_assay <- function(x, dfSE) {
+  dfSE2 <- uTaxaID <- NULL
```
See above

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+    message("Missing taxonomic information")
+    # message("Missing taxonomic information \n returning only molten assay")
+
+    dfSE2 <- dfSE
```
The object does not need to be copied

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+
+    dfSE2 <- dfSE
+    return(dfSE2)
+  } else {
```
The else branch is not necessary

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+      rownames_to_column("uTaxaID")
+    # may need to enforce in case of user supplied duplicate
+
+    dfSE2 <- dfSE %>%
```
duplicated objects not needed

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#' @param add_col_data Choice of colData columns to add. Default="none".
+#'
+.add_col_data_to_molten_assay <- function(dfSE2, x, add_col_data) {
+  dfSE3 <- uSamId <- NULL
```
See above

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+  if (is.null(add_col_data[1]) | is.na(add_col_data[1]) | add_col_data[1] == "none") {
+    message("Returning molten assay without colData")
+
+    dfSE3 <- dfSE2
```
See above

FelixErnst commented on R/utils.R:
```diff
@@ -53,3 +53,18 @@
+# checks
+
+#' @importFrom SummarizedExperiment assays
+.check_abund_values <- function(abund_values, x,
```
Just reimport it like this:

```
.check_abund_values <- MicrobiomeExperiment:::.check_abund_values 
```

Then the function from `MicrobiomeExperiment` becomes a single point of thruth

FelixErnst commented on tests/testthat/test-meltAssay.R:
```diff
@@ -0,0 +1,41 @@
+                              abund_values = "counts")
+    expect_equal(class(molten_assay)[1], "tbl_df")
+    expect_equal(class(molten_assay)[2], "tbl")
+    expect_equal(class(molten_assay)[3], "data.frame")
```
try to use `expect_s3_class`

```
expect_s3_class(tbl, c("tbl_df","tbl","data.frame"))
```

FelixErnst commented on tests/testthat/test-meltAssay.R:
```diff
@@ -0,0 +1,41 @@
+    expect_equal(is.numeric(only_assay$Abundance), TRUE)
+
+    assay_taxa <- .add_taxonomic_data_to_molten_assay(se, only_assay)
+    expect_equal(colnames(assay_taxa)[1], "uTaxaID")
```
`expect_equal(colnames(assay_taxa),c("uTaxaID","uSamId","Abundance","Kingdom"))` is shorter

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#' @export
+setGeneric("meltAssay",
+  signature = "x",
+  function(x, add_col_data = "none", abund_values = "counts", ...) {
```
I think it would be good, if `add_col_data` would accept the following values:

- NULL so no data gets added.
- character values to select certain columns
- TRUE to select all columns.

It could be the case, that a column is named `none` or `all`, which then couldn't be added to the molten data. The same is also true for taxonomic data. I think we need a argument for that as well. Maybe it is also good to use a general term like `add_row_data`. What do you think?

microsud replied:
I made changes to `add_col_data` options. Regarding `add_row_data`, I am not sure if this option will be of use. This is mostly based on my experience with microbiome data. The `rowData` is one of the most useful information for interpretation and data visualization. `phyloseq` also returns `tax_table` by default. Additionally, the hierarchical nature of information with often *missing* classification at various levels hampers interpretation and users would always keep all information at all taxonomic levels.

FelixErnst replied:
I don't see this as a major problem. Maybe this gets us back to a conceptual discussion. `meltAssay` is for me the main function, the workhorse so to say. However, I don't see that this as a strictly user facing functions.

Think about it this way: What task would a user have towards the object and its data? 
1. Give me the abundance data with all/some/these taxonomic information.
2. give me the abundance data with with certain covariate information.

In addition, `melting` is a really technical term, which isn't even used in the tidyverse anymore (they now use `pivot_longer`) and is certainly not a term associated by default with tasks outlined above.

Therefore, at least in my mind, we will have a function `getAbundanceData` or something, which will have different default and will get the task done and use `meltAssay` internally.

So this is why I am strongly arguing, that we keep our options open.

`meltAssay` must be able to add data, but only, when required. If we want to reuse `meltAssay` for some plotting or other internal functions, I don't want to right the function again, just because `meltAssay` gives me data, I didn't ask for. That will blow the memory consumption at the most unfortunate moment, aka. the moment where you are running low on memory. And then this discussion we are having matters.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+
+  molten_assay <- assays(x)[[abund_values]] %>%
+    data.frame() %>%
+    rownames_to_column("uTaxaID") %>%
```
Does the `rownames` need to contain taxonomic information? I think a more general term like `uFeatureID` would be more appropriate and would be consistent with `uSamID` used below.

For what does the `u` stand for in `uTaxaID`? unique?

Maybe `SampleID` and `FeatureID` would do as well, since there are now checks for uniqueness in the code. I think the uniqueness is only derived from `rownames` and `colnames`. In addition the values are not unique in the final molten data.

FelixErnst commented on R/meltAssay.R:
```diff
@@ -0,0 +1,152 @@
+#'
+#' @export
+setMethod("meltAssay",
+  signature = c(x = "TreeSummarizedExperiment"),
```
It does not need to be a `TreeSummarizedExperiment`. Just a `SummarizedExperiment` would work as well, since `SCE` or `TSE` specific features are not used.

FelixErnst reviewed (commented):
Hi Sudarshan @microsud,

The changes are good and certainly in the right direction. Have look through my comments and let me know, what you think. Thanks!

Felix

microsud commented:
Thanks. All good points and will consider them in one more attempt 🙂

microsud committed:
second attempt

FelixErnst reviewed (approved):
It is good to go. The messages are a bit unneccessary, since they will make using meltAssay as an internal function harder

FelixErnst committed:
Merge branch 'master' into meltassay-1

FelixErnst committed:
tweaks for meltAssay

FelixErnst committed:
sync

Merge branch 'meltassay-1' of https://github.com/FelixErnst/mia into meltassay-1

# Conflicts:
#	R/runCCA.R
